peer/peer.go

```diff
@@ -1065,10 +1072,42 @@ func (p *Peer) handlePongMsg(msg *wire.MsgPong) {
 	}
 }
 
-// readMessage reads the next bitcoin message from the peer with logging.
-func (p *Peer) readMessage(encoding wire.MessageEncoding) (wire.Message, []byte, error) {
-	n, msg, buf, err := wire.ReadMessageWithEncodingN(p.conn,
-		p.ProtocolVersion(), p.cfg.ChainParams.Net, encoding)
+// readMessage reads the next bitcoin message from the peer with logging. The
+// partial bool indicates that we've partially read a message already. In this
+// case, we use the ReadPartialMessageWithEncodingN function.
+func (p *Peer) readMessage(encoding wire.MessageEncoding, partial bool) (
+	wire.Message, []byte, error) {
+
+	var (
+		n         int
+		msg       wire.Message
+		buf       []byte
+		plaintext []byte
+		err       error
+	)
+
+	if p.cfg.UsingV2Conn {
+		plaintext, err = p.V2Transport.V2ReceivePacket(nil)
+		if err != nil {
+			fmt.Printf("failed to read packet: %v\n", err)
+			return nil, nil, err
+		}
+
+		msg, buf, err = wire.ReadV2MessageN(
+			plaintext, p.ProtocolVersion(), encoding,
+		)
+		n = len(plaintext)
+	} else if partial {
```

Comment:
?

Reply:
This is used for when we expect a v2 inbound connection but instead receive a v1 version message. In this case, we've already read some of the v1 version header and need to process the rest of it. Without this, the v2->v1 inbound downgrade doesn't work properly.

Reply:
Hmm, ok will examine this relevant area of the BIP further. 

Can we add a comment here explaining the rationale here a bit?